Add *setfavoriteitemidx & *autofavoriteitem script commands - alter item favorite state

Emistry opened the pull request:
<!-- Before you continue, please change "base: stable" to "base: master" and
     enable the setting "[√] Allow edits from maintainers." when creating your
     pull request if you have not already enabled it. -->

<!-- Note: Lines with this <!-- syntax are comments and will not be visible in
     your pull request. You can safely ignore or remove them. -->

### Pull Request Prelude

<!-- Thank you for working on improving Hercules! -->
<!-- Please complete these steps and check the following boxes by putting an `x`
     inside the [brackets] before filing your Pull Request. -->

- [x] I have followed [proper Hercules code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed
- `setfavoriteitemidx()` set an item as favorite item or not
- if an item is set to favorite item, it will be moved into favorite
tab, else move out from favorite tab.
- only non-equipped item can adjust the favorite item state.
- `autofavoriteitem()` will auto set an item as favorite state or not whenever obtained. (affect globally)

-----------------------------

<!-- Describe the changes that this pull request makes. -->
Sample Script: `autofavoriteitem(<itemID>, <flag>)`
<details>

```
prontera,155,181,4	script	Sample	4_F_KAFRA1,{
	.@item_id = 909;
	mes getitemname(.@item_id);
	// mes "auto favorite: "+getiteminfo2(.@item_id, 21);
	.@flag = select("disable favorite", "enable favorite") - 1;
	autofavoriteitem(.@item_id, .@flag);
	close;
}
```
</details>

Sample Script: `setfavoriteitemidx(<idx>, <flag>)`
<details>

```
prontera,155,181,4	script	Sample	4_F_KAFRA1,{
	do {
		getinventorylist;
		.@menu$ = "";
		for (.@i = 0; .@i < @inventorylist_count; .@i++) {
			.@c = (@inventorylist_card1[.@i] > 0) + (@inventorylist_card2[.@i] > 0) + (@inventorylist_card3[.@i] > 0) + (@inventorylist_card4[.@i] > 0);
			.@menu$ += "[idx- "+@inventorylist_idx[.@i]+"] " + getitemname(@inventorylist_id[.@i]);
			if (.@c) 
				.@menu$ += "["+.@c+" cards]";
			.@menu$ += " "+ (@inventorylist_favorite[.@i] ? "^0055FF- favorite^000000":"") + ":";
		}
		.@i = select(.@menu$) - 1;
		// based on inventory index
		setfavoriteitemidx(@inventorylist_idx[.@i], !@inventorylist_favorite[.@i]);
	} while (true);
	close;
}
```

`@inventorylist_favorite[.@i]` [from this PR](https://github.com/HerculesWS/Hercules/pull/2426), but not really dependant on it. Since it can be just set with the value `0 or 1`. Bonus if have it.
</details>

**Issues addressed:** <!-- Write here the issue number, if any. -->
None

<!-- You can safely ignore the links below:  -->

[cont]: https://github.com/HerculesWS/Hercules/blob/master/CONTRIBUTING.md
[code]: https://github.com/HerculesWS/Hercules/wiki/Coding-Style

hemagx commented on src/map/script.c:
```diff
@@ -24559,6 +24559,28 @@ static BUILDIN(getcalendartime)
+		ShowError("buildin_setfavoriteitem: no player attached.\n");
+		return false;
+	}
+	else if (idx < 0 || idx >= sd->status.inventorySize || sd->status.inventorySize <= 0) {
```
``` } else if (x) { ```

hemagx replied:
Also there's no need to check the inventory size, the index check is enough to validate it

Emistry replied:
done

hemagx commented on src/map/script.c:
```diff
@@ -24559,6 +24559,28 @@ static BUILDIN(getcalendartime)
+		ShowError("buildin_setfavoriteitem: Invalid Inventory Index %d (min: %d, max: %d).\n", idx, 0, (sd->status.inventorySize - 1));
+		return false;
+	}
+	else if (sd->status.inventory[idx].equip == 0) {
```
``` } else if (x) { ```

Emistry replied:
done

dastgirp commented on src/map/script.c:
```diff
@@ -24559,6 +24559,26 @@ static BUILDIN(getcalendartime)
+	if (sd == NULL) {
+		ShowError("buildin_setfavoriteitem: no player attached.\n");
+		return false;
+	} else if (idx < 0 || sd->status.inventorySize <= 0) {
```
check "idx" with max size?
because this may cause crash if idx is not proper.
e.g:
```
idx = 100
sd->status.inventorySize = 60
condition: if (100 < 0 || 60 < 0) => False
then it would go below at line 24574 and crash.

Emistry replied:
done

AnnieRuru commented on src/map/script.c:
```diff
@@ -24559,6 +24559,26 @@ static BUILDIN(getcalendartime)
+	if (sd == NULL) {
+		ShowError("buildin_setfavoriteitem: no player attached.\n");
+		return false;
+	} else if (idx < 0 || sd->status.inventorySize <= 0 || idx >= sd->status.inventorySize) {
```
what's the point of having `sd->status.inventorySize <= 0` ?
isn't this always between 100 and MAX_INVENTORY ?

Emistry replied:
checking condition for if player has no item in inventory

AnnieRuru replied:
no, `sd->status.inventorySize` is equivalent to `*getInventorySize` script command
`sd->status.inventorySize` is the maximum amount of different items the player can hold

for example, if I have 10 items inside my inventory, 
`sd->status.inventorySize` is still 100, and can only be expand with `*expandInventory` script command

Emistry replied:
i see. thx for the info. updated it.

AnnieRuru commented:
btw, why using inventory index for this ?
https://github.com/AnnieRuru/Release/blob/master/plugins/favorite_tab/favorite_tab_0.1.c
I did it with ItemID instead of inventory index ...

hmm ... so ... 
using ItemID can ... when pick up this ItemID, can immediately move to favorite tab
using inventory index ... when there are multiple ItemID ...
-> example equipment with same ID, only that equipment will move to favorite tab ...
hmmm ....

AnnieRuru commented:
how about split into 3 script commands
similar to `*equip` `*equipidx` `*autoequip`
so there will be `*setfavoriteitem` `*setfavoriteitemidx` and `*autofavoriteitem` ?

Emistry commented:
for itemID as parameter,
it doesn't take the correct/desired item if the character has multiple same itemID or item with different refine/itembonus/cards/etc _(I am sure you had experienced same issue in the past using the `autoequip`, delitem`, and etc)_
using `getinventorylist` could have avoided all that since all info of that item can be obtained and check during the script execution, it even pick the correct and actual inventory index that you intended to set to favorite state.
its possible to implement with itemID/index as parameter, but dont really see the needs for it.
for multiple new script command for the purpose of separating parameter index or itemid? I did vote for -1 for this.
imo, using itemID as parameter are less flexible/convenient than inventory index.

AnnieRuru commented:
> its possible to implement with itemID/index as parameter, but dont really see the needs for it.
for multiple new script command for the purpose of separating parameter index or itemid? I did vote for -1 for this.
imo, using itemID as parameter are less flexible/convenient than inventory index.

com'on, you didn't read the topic link in rathena forum
https://rathena.org/board/topic/118227-consumables-in-favorites/
Seravy-sensei wants the blue-gemstone in the favorite tab forever
this can only done with `*autofovariteitem`, set 717 as permanent item in favorite tab

... even if I said I can always release as plugin on the forum,
I also wish this script command name as `*setfavoriteitemidx` instead
to avoid confusion with my release later,
as your script command clearly only move item by inventory index

Emistry commented:
`setfavoriteitem` I have added support for item_id, if the value provided are higher than inventory size, it will be treated as item_id.
But still, I don't recommend using itemID since its lack of accuracy to determine the item in case player has multiple same item id in their inventory.

`autofavourite` I have added this to auto set the item as favorite state if player obtained it.
This is actually a good idea to implement it. The use-case are way common if compared to `setfavoriteitem`.

Kindly help re-review the PR. 
Thank you in advance.

Emistry commented:
the patch check for `if (nameid >= ITEMID_RED_POTION || nameid >= sd->status.inventorySize) {`.
this is also one of the reason I don't wish it support item ID, since its silly whenever a player try to create an item that shouldn't fall below "minimum" item id, and gravity is brilliant enough to start itemid at 501. lol


btw, kindly  dont use "request changes" to add a comment without really need anything to changes in the code? lol

AnnieRuru commented:
> gravity is brilliant enough to start itemid at 501

LOL !! I actually want to say gravity is stupid enough to not start the ID from 1 or 0

> btw, kindly dont use "request changes" to add a comment without really need anything to changes in the code? lol

yes, because we can use ~item_avail~ `Sprite` field to create itemID gravity not using
1-500 fall within the possible range that any server can use these ID range
means your code actually create conflicts for those server that actually uses these ID range

Emistry commented:
I am leaving the restriction here since there are no official way to tell the `min_item_id` that a server is using, unless they self-defined the MIN_ITEM_ID or something equivalent. 
Not to mention the current `max_item_id` support up to millions, I dont really see a point using `itemID <= 500` as their custom item id.

if I to changed the code to give priority for legit `item_id` checking, it would lost the function to alter the state based on inventory index if an item ID exists with the same number as the inventory index. 
Which at the end, resulting to force creating another equivalent script command to support it.
<details>

```
	// if item id are provided
	if (id != NULL) {
		idx = pc->search_inventory(sd, nameid);
		if (idx == INDEX_NOT_FOUND) {
			ShowError("buildin_setfavoriteitem: Item %d not found.\n", nameid);
			return false;
		}
	}
```
</details>

this is why itemID as parameter aren't preferable in the beginning, it has lack accuracy, and create unnecessary works, which originally could be avoided if they use `getinventolist()` to get the `inventory_idx` value to use with this script command.

AnnieRuru commented:
> Which at the end, resulting to force creating another equivalent script command to support it.

I have been saying separate into `*setfavoriteitem` and `*setfavoriteitemidx`
if your idea actually accurate, that also means the PR about [*delitemidx](https://github.com/HerculesWS/Hercules/pull/2394) and [*equipidx](https://github.com/HerculesWS/Hercules/pull/2355) shouldn't even exist
and should alter existing `*delitem` and `*equip` script command instead, which create further confusion

ok now wait 1 month until Haru come back and review this

MishimaHaruna commented:
I also think that having a command that can take either an index or a nameid is a recipe for problems.

My main concerns are:
- A customized setting, or a future Gravity update could push the minimum item ID to < 500
- A customized setting/client or a future Gravity update could push the maximum inventory size to > 500
- A malfunctioning script that expects to have an item ID in a variable could have the wrong value (or it could have a typo in a variable or constant, resulting into a `0`), which could be silently accepted by the script as an inventory index, instead of throwing a runtime error and aborting the script because of an invalid nameid.

I'd suggest to accept either the nameid or the index (and have separate commands if you want to accept both). It's not nice to have several commands that do the same thing, but it seems to me the approach that's less prone to errors.

Emistry commented:
Then I will just proceed with original plan, only accept `inventory_index`.
imo, using itemID as parameter are less flexible/convenient than inventory index.

Asheraf commented:
Following up on what Haru daid:
   - Gravity already have an item with id = 1
```
	[1] = {
		unidentifiedDisplayName = "머리띠",
		unidentifiedResourceName = "눈가리개",
		unidentifiedDescriptionName = {},
		identifiedDisplayName = "마력의 별",
		identifiedResourceName = "마력의별",
		identifiedDescriptionName = {},
		slotCount = 0,
		ClassNum = 0
	},
```

- Inventory Expansion feature allows for > 500 inventory size (herc still have some packet limits but client support it normally)
